core/storage/base_model/gae_models.py

```diff
@@ -85,7 +85,8 @@ class BaseModel(ndb.Model):
 
     # When this entity was first created. This can be overwritten and
     # set explicitly.
-    created_on = ndb.DateTimeProperty(indexed=True, required=True)
+    created_on = ndb.DateTimeProperty(
```

Comment:
@vojtechjelinek I think you deleted auto_now_add some time ago, in PR #7625. Could you PTAL at this and check that the change makes sense?

Also -- was it a mistake? It seems to me like it might have been, given the issues that we are seeing that this PR aims to fix.

Reply:
Hmm, I think it was wrong understanding of how we handle models on the domain level, it seems that we quite frequently do not copy the `created_on` field through the domain level that means when the model is saved it has `created_on` set to `None` and it gets updated (see line 250 of this file).

Reply:
I assume that the `auto_now_add` is more complex and checks not only if `created_on` is `None` in the model instance, but also if the model with the `id` equal to the instance `id` has `created_on` set on.

Reply:
@iamprayush Could you please check this? If an already existing model with some `id` is put into the datastore with `created_on` set to `None` (in the datastore `created_on` is set to some time before) what is the value of `created_on` after the `put`?

Reply:
@vojtechjelinek I checked and a `put` request updates the `created_on` field from `None` to the time at which the `put` request was made and after that, it stays as it is even upon making more put requests. And if I'm not wrong, that is the expected behavior.

Reply:
Even more put requests when `created_on` is again set to `None`?

Reply:
No, I mean, once the `created_on` has been updated from `None` to the time of the put request, it then doesn't change after that.

Reply:
But I'm asking if `auto_now_add` works in a different way than our own `put` method already does (see line 250 of this file).

Because if it works the same way we will need to audit all the other places in our code and see if we always pass `created_on` on the domain level and never set it to `None`. So the created on is not auto added after the model was existing. And we don't need to add this.

If it works in a way that even if `created_on` is set to `None` in our code it keeps the old date (since the old date is saved in the datastore) then it makes sense to add this.

Reply:
I'm sorry I don't understand what you mean by this:
> If it works in a way that even if created_on is set to None in our code it keeps the old date (since the old date is saved in the datastore) then it makes sense to add this.

Isn't this the expected behavior from either of the implementations (auto_now_add or the one on line 250)? I mean, the value of `created_on` _is_ set to `None` in our code in both cases, right?  
I think the reason that the `auto_now` & `auto_now_add` parameters were not being used till now was that we don't always want the `last_updated_on` field to be updated which is why the put method takes in a boolean parameter for it. This is why I haven't used `auto_now` for `last_updated_on` but we _can_ use `auto_now_add` for `created_on` since there are no restrictions on updating that field.

And to answer your question: According to the [documentation](https://cloud.google.com/appengine/docs/standard/python/ndb/entity-property-reference) and based on my observations locally, I can confirm that `auto_now_add` works in the same way as our current implementation.

Reply:
> Isn't this the expected behavior from either of the implementations (auto_now_add or the one on line 250)? I mean, the value of created_on is set to None in our code in both cases, right?

Maybe, I think it makes sense to have it this way but it also would makes sense that the model would be crosschecked with datastore if it isn't already created and the value wouldn't be updated.

Then we don't need to add `auto_now_add` since we already have it in our code? And we need to either check our codebase that we always preserve `created_on` OR implement our `put` in a way that if `created_on` is set on an existing model to `None` we will use the value that is in the datastore and not overwrite it with the current time.

Reply:
Or we could remove the updating `created_on` functionality from line 250 and just rely on `auto_now_add`? That seems to make the code simpler, doesn't it?
Also, with the current implementation, the value of `created_on` will be preserved since we're always checking for the existence of the model in the _save_user_settings method.

Reply:
Yep, but we still need to do the audit for the correct handling of `created_on` in other models.

Reply:
I've updated this PR.
Also, I checked the other models and I see that they (exploration, collection, question, skill, story, and topic) all have their "custom" `created_on` attributes. For instance, the collection model has a `collection_model_created_on` attribute which is then persisted while creating or modifying instances of the model.

Reply:
**Update**: I've used the normal implementation of updating the `created_on` field since using `auto_now_add` for `created_on` and using normal implementation for `last_updated_on` causes some issues. The problem caused is that `created_on` gets updated before `last_updated_on` which is not what we want. This happens because `created_on` is being updated by a property that's inherent to the DateTime object whereas `last_updated_on` is being initialized and updated externally by put method.
Also, just to confirm, both the methods (with and without `auto_now_add`) have the same effect on the data, so there should be no more issues now.